models/particles/d3q27_PSM/Dynamics.c.Rt

```diff
@@ -0,0 +1,1067 @@
+    
+    <?R 
+    fpop = PV(paste("f",1:27-1,"(adjX,0,0)",sep=""))
+    d=PV("d") 
+    C(d, sum(fpop)) 
+    
+
+    ZouHe(EQ,1,-1,"pressure")
+
+    ?>
+}
+
+CudaDeviceFunction void PressureExitW()
+{
+    real_t u[3], d, eq_local[27], eq_neigh[27], usq;
+    const int adjX = 1; 
+    
+    <?R 
+    fpop = PV(paste("f",1:27-1,"(adjX,0,0)",sep=""))
+    d=PV("d") 
+    C(d, sum(fpop)) 
+    
+
+    ZouHe(EQ,1,1,"pressure")
+
+    ?>
+}
+
+CudaDeviceFunction void PressureExitF()
+{
+    real_t u[3], d, eq_local[27], eq_neigh[27], usq;
+    const int adjZ = -1; 
+    
+    <?R 
+    fpop = PV(paste("f",1:27-1,"(0,0,adjZ)",sep=""))
+    d=PV("d") 
+    C(d, sum(fpop)) 
+    
+
+    ZouHe(EQ,3,-1,"pressure")
+
+    ?>
+}
+
+CudaDeviceFunction void PressureExitB()
+{
+    real_t u[3], d, eq_local[27], eq_neigh[27], usq;
+    const int adjZ = 1; 
+    
+    <?R 
+    fpop = PV(paste("f",1:27-1,"(0,0,adjZ)",sep=""))
+    d=PV("d") 
+    C(d, sum(fpop)) 
+    
+
+    ZouHe(EQ,3,1,"pressure")
+
+    ?>
+}
+
+CudaDeviceFunction void MovingNWall()
+{
+        f[1]  = f[3];
+        f[16] = f[17];
+        f[18] = f[15];
+
+        f[21]  = f[24] + VelocityX/36.0;
+        f[25] = f[20]  + VelocityX/36.0;
+        f[9] = f[8] + VelocityX/9.0;
+
+        f[22] = f[23] - VelocityX/36.0;
+        f[26]= f[19]  - VelocityX/36.0;
+        f[10] = f[7] - VelocityX/9.0;
+}
+
+CudaDeviceFunction void MovingSWall(){
+        f[3]  = f[4];
+        f[17] = f[16];
+        f[15] = f[18];
+
+        f[24] = f[21]  - VelocityX/36.0;
+        f[20]  = f[25] - VelocityX/36.0;
+        f[8] = f[9] - VelocityX/9.0;
+
+        f[23] = f[22] + VelocityX/36.0;
+        f[19]  = f[26] + VelocityX/36.0;
+        f[7] = f[10] + VelocityX/9.0;
+}
+
+
+CudaDeviceFunction void SetEquilibrum(const real_t d, const real_t u[3])
+{	real_t usq, uf;
+<?R C(usq, sum(u*u)*(1/c_squ)) ?>
+<?R
+	for (i in 1:27) {
+         cat("//--",i,"-------------------------------------------------\n")
+         C( uf1, (u %*% t(U) * (1/c_squ))[i])
+         C( uf2, uf1 + (uf1*uf1+usq*(-1))*0.5 + 1)
+         C( uf3, d * uf2)
+         C( f[i], uf3 * tn[i,i])
+        }
+?>
+
+}
+
+CudaDeviceFunction void SetEquilibrum2(real_t rho, real_t Jx, real_t Jy, real_t Jz)
+{
+	<?R
+		C(f, EQ$Req %*% solve(EQ$mat));
+	?>
+}
+
+CudaDeviceFunction real_t CalcSolid(){
+        real_t coverage = 0, ret = 0;
+
+        for (ParticleIterator p(X,Y,Z); p; ++p) if ((p.dist - p.rad)<1.5) {
+                if ((p.dist - p.rad) < - 1){
+                        coverage = 1.0;
+                } else{
+                        coverage = (p.rad -  0.084/p.rad + 0.5 - p.dist);
+                }
+
+                if (coverage > 1.0){ coverage =  1.0;}
+                if (coverage < 0.0){ coverage =  0.0;}
+
+                ret += coverage;
+
+        }
+
+        if (ret > 1.0){ ret =  1.0;}
+        if (ret < 0.0){ ret = 0.0;}
+
+        return ret;
+}
+
+CudaDeviceFunction real_t CalcPeriodicSolid(){
```

Comment:
This is obsolete (the lammps integration has periodicity, and ESYS will have) and inefficient (loops over all particles). I think you can get rid of it.